maven-di/src/main/java/org/apache/maven/di/Injector.java

```diff
@@ -0,0 +1,613 @@
+ */
+package org.apache.maven.di;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.*;
+import java.util.Map.Entry;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+import java.util.function.Supplier;
+import java.util.function.UnaryOperator;
+
+import org.apache.maven.api.annotations.Nullable;
+import org.apache.maven.di.binding.*;
+import org.apache.maven.di.impl.CompiledBinding;
+import org.apache.maven.di.impl.CompiledBindingLocator;
+import org.apache.maven.di.impl.Preprocessor;
+import org.apache.maven.di.module.Module;
+import org.apache.maven.di.module.Modules;
+import org.apache.maven.di.util.Trie;
+import org.apache.maven.di.util.Utils;
+
+import static java.util.stream.Collectors.toMap;
+import static org.apache.maven.di.Scope.UNSCOPED;
+import static org.apache.maven.di.util.Types.parameterizedType;
+
+/**
+ * Injector is the main working component of the ActiveJ Inject.
+ * <p>
+ * It stores a trie of binding graphs and a cache of already made singletons.
+ * <p>
+ * Each injector is associated with exactly zero or one instance per {@link Key}.
+ * <p>
+ * Injector uses binding graph at the root of the trie to recursively create and then store instances of objects
+ * associated with some {@link Key keys}.
+ * Branches of the trie are used to {@link #enterScope enter scopes}.
+ */
+@SuppressWarnings({"unused", "WeakerAccess", "rawtypes"})
+public final class Injector implements ResourceLocator {
+    static class ScopeLocalData {
+        final Scope[] scope;
+        final Map<Key<?>, Binding<?>> bindings;
+        final Map<Key<?>, CompiledBinding<?>> compiledBindings;
+        final Map<Key<?>, Integer> slotMapping;
+        final int slots;
+        final CompiledBinding<?>[] eagerSingletons;
+
+        ScopeLocalData(
+                Scope[] scope,
+                Map<Key<?>, Binding<?>> bindings,
+                Map<Key<?>, CompiledBinding<?>> compiledBindings,
+                Map<Key<?>, Integer> slotMapping,
+                int slots,
+                CompiledBinding<?>[] eagerSingletons) {
+            this.scope = scope;
+            this.bindings = bindings;
+            this.compiledBindings = compiledBindings;
+            this.slotMapping = slotMapping;
+            this.slots = slots;
+            this.eagerSingletons = eagerSingletons;
+        }
+    }
+
+    final @Nullable Injector parent;
+    final Trie<Scope, ScopeLocalData> scopeDataTree;
+
+    final Map<Key<?>, Integer> localSlotMapping;
+    final Map<Key<?>, CompiledBinding<?>> localCompiledBindings;
+
+    final AtomicReferenceArray[] scopeCaches;
+
+    private static Supplier<UnaryOperator<CompiledBinding<?>>> bytecodePostprocessorFactory = UnaryOperator::identity;
+
+    /**
+     * Enables specialization of compiled bindings. <b>Depends on {@code ActiveJ-Specializer} module</b>
+     */
+    @SuppressWarnings("unchecked")
+    public static void useSpecializer() {
```

Comment:
do we use it? if so we should either require it (no reflection) or drop it no?

Reply:
I don't think we need, so that will be removed.